lib/livebook/runtime/evaluator.ex

```diff
@@ -688,20 +692,100 @@ defmodule Livebook.Runtime.Evaluator do
     {result, code_markers}
   end
 
-  defp eval(:erlang, code, binding, env) do
+  # Erlang code is either statements as currently supported, or modules.
+  # In case we want to support modules - it makes sense to allow users to use
+  # includes, defines and thus we use the epp-module first - try to find out
+  #
+  # if in the tokens from erl_scan we find at least 1 module-token we assume
+  # that the user is defining a module, if not the previous code is called.
+
+  defp eval(:erlang, code, binding, env, tmp_dir) do
+    case :erl_scan.string(String.to_charlist(code), {1, 1}, [:text]) do
+      {:ok, tokens, _} ->
+        case :lists.keyfind(:module, 3, tokens) do
+          {:atom, _, _module} -> eval_erlang_module(code, binding, env, tmp_dir)
+          false -> eval_erlang_statements(code, tokens, binding, env)
+        end
+
+      {:error, {location, module, description}, _end_loc} ->
+        process_erlang_error(env, code, location, module, description)
+    end
+  end
+
+  # Explain to user: without tmp_dir to write files, they cannot compile erlang-modules
+  defp eval_erlang_module(_code, _binding, _env, nil) do
+    {{:error, :error, "erlang module compile needs tmp_dir", []}, []}
+  end
+
+  # Create module - tokens from string
+  # Based on: https://stackoverflow.com/questions/2160660/how-to-compile-erlang-code-loaded-into-a-string
+  # Step 1: Scan the code using the epp:parse_file erlang primitive
+  # - epp will do macro expansion, includes, defines etc.
+  # Step 2: Compile and load
+  # Step 3: If compile success - register module
+
+  defp eval_erlang_module(code, binding, env, tmp_dir) do
+    # Consider using in-memory file, once :ram file supports IO device API.
+    # See https://github.com/erlang/otp/issues/7239
+    filename = Path.join(tmp_dir, "epp#{random_long_id()}.tmp")
+    File.write!(filename, code)
+
+    eval_result =
+      case :epp.parse_file(~c"#{filename}", [], []) do
+        {:ok, forms} ->
+          case :lists.keyfind(:module, 3, forms) do
+            {:attribute, _lineno, :module, module} ->
```

Comment:
We already checked for module before, so there is no need to check here again, right?

Reply:
In the first check I only check if a module-attribute exists.
I tried using epp for only erlang-statements but it did not return usable information.

Since the entire conversion from code to forms is done by epp and now we deduce in that step if we have a file-system it seems to be a simpler logic without having to pass tokens around.